### DISCUSSION TITLE
插入参考文献自动补全 citation key 的方法

### DISCUSSION THREAD

**Discussion by dreistein543** [nju-lug/NJUThesis#126] [Show and tell]
目前模板默认使用 `biblatex` 处理参考文献，但是，_只少在 mac 环境下_，使用 texstudio 无法在使用 `\cite{}` 时自动补全 citation key，这大大降低了我们写论文的流畅性。

> ~~之前，我采用 在文末 添加 `bibtex` 的 `\bibliography{}` 来完成自动补全，**但这需要在编译的时候去掉，一个不留神会导致编译失败。**~~

**现在我发现，其实只要使用 `\addbibresource{}` 替换掉 `resource = {}`，就可以达成目的**

```latex
	bib = {
		style = author-year,
		% 参考文献样式
		% 默认为顺序编码制（numeric）
		% 可选著者-出版年制（author-year）
		%
		% resource = {bib文件路径},               % 通过注释该行，停止使用 resource = {}
		% 参考文献数据源
		% 需要带扩展名的完整文件名
		% 可使用逗号分隔多个文件
		% 此条等效于 \addbibresource 命令
		%
		% option = {
			% doi    = false,
			% isbn   = false,
			% url    = false,
			% eprint = false,
			% 关闭部分无用文献信息
			%
			% refsection = chapter
			% 将参考文献表置于每章后
			%   }
		% 额外的 biblatex 宏包选项
	},

\addbibresource{bib文件路径},          % 直接在这里指定 bib 文件就好，但是一定要在 bib = {}, 的外面，否则编译会报错
```

如此，就可以做到键入 `\cite{}` 时，自动补全 `bib` 文件里文献的 citation key 了。

**Comment by atxy-blip** [comment]
实际上应该在 `\njusetup` 命令的外面。`\njusetup` 的参数只能是一个键值对列表。

**Reply by dreistein543** [reply]
我一开始原地替换导致编译错误，后来就放在`\begin{document}`之前就好了，有大佬解答具体情况真是帮了大忙了

**Comment by Liuyulun** [comment]
> 实际上应该在 `\njusetup` 命令的外面。`\njusetup` 的参数只能是一个键值对列表。

请问这是什么意思？我已用申明了\addbibresource{References.bib}，还是不能自动补全ref。不过我使用的是NKU的模板

**Reply by dreistein543** [reply]
没看过NKU模板，这里的意思是指，在原模板设定参数的部份之外额外进行设置。可以简单理解为，在`\begin{document}` 之前一行单独添加（不能包含在任何其他命令的花括号里）

**Reply by tangruize** [reply]
我之前也遇到过这个问题（https://github.com/nju-lug/NJUThesis/issues/288 ）。实际上要让自动补全正常work，bib 里的条目还不能包含特殊字符，比如 “@”。如果是这个问题造成的，编译器 warning 也会给提示。希望这个经验能帮到更多人。

**Comment by Liuyulun** [comment]
按照友友的方法，试了一下还是不行呀

---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 发送日期 | 2024年08月27日 00:19 |
| 收件人 | nju-lug/NJUThesis ***@***.***> |
| 抄送人 | Liuyulun ***@***.***>,
Comment ***@***.***> |
| 主题 | Re: [nju-lug/NJUThesis] 插入参考文献自动补全 citation key 的方法 (Discussion #126) |

没看过NKU模板，这里的意思是指，在原模板设定参数的部份之外额外进行设置。可以简单理解为，在\begin{documents} 之前一行单独添加（不能包含在任何其他命令的花括号里）

**Reply by dreistein543** [reply]
我觉得你在询问之前应该进一步明确自己的情况，比如你是用的什么什么方式（在线编译还是本地编译？在线是什么平台？本地是什么编辑器？LaTeX版本是什么？模板的版本是什么？以及最好是附赠一个你当前的可以运行的最小代码（包含了tex,bib,def,sty以确保可以输出正确PDF的代码或文件，当然自行模糊掉个人信息和论文信息）供他人复现测试）

最后就是，既然你用的NKU模板，优先去NKU模板的页面反馈问题可能能够获得更好的回答与支持

**Comment by Liuyulun** [comment]
谢谢友友，我在NKU的issue的下面提问了，但是没人回答，然后看到相同的问题就问留言了，没想到你答复我了。我用texstudio编译可以通过，但是\cite不能显示ref。今天试了一下在vs code中编译不能通过，但是能在\cite中显示bib文献。

下面展示texstudio编辑器中的情形。bib中citation key无误，但是无法自动补全。我键入【yuan】就显示【yuan】。texstudio中自动补齐功能也打开了。
NKU模板以附件形式发送。
非常感谢你的回复，我知道你本来不需要这样做。如果您有时间，希望您能看看是哪里有问题。再次感谢！












在 2024-08-27 15:02:57，"dreistein543" ***@***.***> 写道：

我觉得你在询问之前应该进一步明确自己的情况，比如你是用的什么什么方式（在线编译还是本地编译？在线是什么平台？本地是什么编辑器？LaTeX版本是什么？模板的版本是什么？以及最好是附赠一个你当前的可以运行的最小代码（包含了tex,bib,def,sty以确保可以输出正确PDF的代码或文件，当然自行模糊掉个人信息和论文信息）供他人复现测试）

最后就是，既然你用的NKU模板，优先去NKU模板的页面反馈问题可能能够获得更好的回答与支持
